File: android/app/src/main/java/com/laneshadow/ui/components/molecules/IntentSearchSheet.kt
package com.laneshadow.ui.components.molecules

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.laneshadow.ui.components.atoms.IconSymbol
import com.laneshadow.theme.LocalLaneShadowTheme

/**
 * Search state sealed class
 *
 * Each state represents a distinct visual treatment for the intent search sheet.
 */
sealed class SearchState {
    /** Empty input field ready for user input */
    object Idle : SearchState()

    /** Instant results with no spinner, shows intent summary pill */
    data class CacheHit(val summary: String) : SearchState()

    /** Loading spinner with status message */
    object Searching : SearchState()

    /** "Connect to search" message with recent-intent chips */
    data class OfflineUnsupported(val recentIntents: List<String>) : SearchState()

    /** Shows intent summary pill with results */
    data class Results(val summary: String) : SearchState()
}

/**
 * IntentSearchSheet molecule component
 *
 * Bottom sheet for natural language route discovery (UC-DISC-07).
 * Displays four visual states: idle, cache_hit, searching, offline_unsupported.
 * Following React Native wrapper patterns from react-native/components/discovery/intent-search-sheet.tsx
 *
 * States:
 * - idle: Empty input field ready for user input
 * - cache_hit: Instant results with no spinner, shows intent summary pill
 * - searching: Loading spinner with status message
 * - offline_unsupported: "Connect to search" message with recent-intent chips
 * - results: Shows intent summary pill
 *
 * @param searchState Current search state — determines which UI to render
 * @param onSearch Callback when user submits search query
 * @param onClear Callback when user taps clear button (resets to browse mode)
 * @param onRecentIntentTap Callback when user taps a recent-intent chip in offline state
 * @param visible Whether the sheet is visible
 * @param query Current search query text
 * @param onChangeQuery Callback when query text changes
 * @param modifier Modifier for the sheet container
 * @param testID Optional test identifier for UI testing
 */
@Composable
fun IntentSearchSheet(
    searchState: SearchState,
    onSearch: (String) -> Unit,
    onClear: () -> Unit,
    onRecentIntentTap: (String) -> Unit,
    visible: Boolean,
    query: String,
    onChangeQuery: (String) -> Unit,
    modifier: Modifier = Modifier,
    testID: String? = null,
) {
    if (!visible) {
        return
    }

    val theme = LocalLaneShadowTheme.current
    val isSearching = searchState == SearchState.Searching

    // Build accessibility description
    val accessibilityDescription = when (searchState) {
        is SearchState.Idle -> "Describe your ideal ride. Enter your search query."
        is SearchState.CacheHit -> "Cache hit: ${searchState.summary}"
        is SearchState.Searching -> "Searching for your perfect ride"
        is SearchState.OfflineUnsupported -> "You're offline. Connect to search or try a recent ride."
        is SearchState.Results -> "Results: ${searchState.summary}"
    }

    AlertDialog(
        onDismissRequest = onClear,
        modifier = modifier
            .testTag(testID ?: "intent-search-sheet")
            .semantics {
                contentDescription = accessibilityDescription
            },
        title = null,
        text = {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = theme.space.md),
                verticalArrangement = Arrangement.spacedBy(theme.space.md),
            ) {
                // Header with title and subtitle
                Column(
                    verticalArrangement = Arrangement.spacedBy(theme.space.xs),
                ) {
                    // Title: "Describe your ideal ride" (title.lg)
                    Text(
                        text = "Describe your ideal ride",
                        color = theme.colors.onSurface.default,
                        style = theme.type.title.lg,
                        modifier = Modifier
                            .fillMaxWidth()
                            .testTag(testID?.let { "$it-title" } ?: "intent-search-title"),
                    )

                    // Subtitle: "Say \"scenic coastal roads\" or \"twisty mountain passes\"" (body.md)
                    Text(
                        text = "Say \"scenic coastal roads\" or \"twisty mountain passes\"",
                        color = theme.colors.onSurface.default.copy(alpha = 0.6f),
                        style = theme.type.body.md,
                        modifier = Modifier
                            .fillMaxWidth()
                            .testTag(testID?.let { "$it-subtitle" } ?: "intent-search-subtitle"),
                    )
                }

                // Divider
                Spacer(
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(1.dp)
                        .background(theme.colors.border.default)
                        .testTag(testID?.let { "$it-divider" } ?: "intent-search-divider"),
                )

                // Input Row with Search Icon, Text Field, and Clear Button
                Surface(
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(48.dp)
                        .testTag(testID?.let { "$it-input-row" } ?: "intent-search-input-row"),
                    shape = RoundedCornerShape(theme.radius.lg),
                    color = theme.colors.surface.default,
                    border = BorderStroke(1.dp, theme.colors.border.default),
                ) {
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(horizontal = 12.dp),
                        horizontalArrangement = Arrangement.spacedBy(theme.space.sm),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        // Search icon (magnify, 20dp)
                        IconSymbol(
                            name = "search",
                            size = 20.dp,
                            color = theme.colors.onSurface.default.copy(alpha = 0.6f),
                            modifier = Modifier
                                .padding(start = 4.dp)
                                .testTag(testID?.let { "$it-search-icon" } ?: "intent-search-icon"),
                        )

                        // Text input field
                        BasicTextField(
                            value = query,
                            onValueChange = onChangeQuery,
                            modifier = Modifier
                                .weight(1f)
                                .testTag(testID?.let { "$it-input" } ?: "intent-search-input"),
                            enabled = !isSearching,
                            singleLine = true,
                            textStyle = TextStyle(
                                color = theme.colors.onSurface.default,
                                fontSize = theme.type.body.lg.fontSize,
                                fontWeight = theme.type.body.lg.fontWeight,
                            ),
                            cursorBrush = SolidColor(theme.colors.primary.default),
                            keyboardOptions = KeyboardOptions.Default.copy(
                                keyboardType = KeyboardType.Text
                            ),
                            decorationBox = { innerTextField ->
                                if (query.isEmpty()) {
                                    Text(
                                        text = "Type your ideal ride...",
                                        color = theme.colors.onSurface.default.copy(alpha = 0.5f),
                                        style = theme.type.body.lg,
                                    )
                                }
                                innerTextField()
                            },
                        )

                        // Clear button (close icon)
                        IconButton(
                            onClick = onClear,
                            enabled = query.isNotEmpty() || isSearching,
                            modifier = Modifier
                                .size(32.dp)
                                .testTag(testID?.let { "$it-clear-button" } ?: "intent-search-clear-button"),
                        ) {
                            IconSymbol(
                                name = "close",
                                size = 20.dp,
                                color = theme.colors.onSurface.default.copy(alpha = 0.6f),
                            )
                        }
                    }
                }

                // State-specific content
                when (searchState) {
                    is SearchState.Idle -> {
                        // No additional content for idle state
                        Spacer(modifier = Modifier.height(theme.space.md))
                    }

                    is SearchState.CacheHit -> {
                        // Cache hit: show intent summary pill
                        IntentSummaryPill(
                            summary = searchState.summary,
                            onDismiss = onClear,
                            modifier = Modifier
                                .testTag(testID?.let { "$it-cache-hit" } ?: "intent-search-cache-hit"),
                        )
                    }

                    is SearchState.Searching -> {
                        // Searching: spinner with status message
                        Column(
                            modifier = Modifier
                                .fillMaxWidth()
                                .padding(vertical = 32.dp)
                                .testTag(testID?.let { "$it-searching" } ?: "intent-search-searching"),
                            horizontalAlignment = Alignment.CenterHorizontally,
                            verticalArrangement = Arrangement.spacedBy(theme.space.md),
                        ) {
                            // Activity indicator spinner
                            CircularProgressIndicator(
                                modifier = Modifier
                                    .size(theme.space.xl)
                                    .testTag(testID?.let { "$it-spinner" } ?: "intent-search-spinner"),
                                color = theme.colors.primary.default,
                            )

                            // "Finding your perfect ride..." (body.lg)
                            Text(
                                text = "Finding your perfect ride...",
                                color = theme.colors.onSurface.default,
                                style = theme.type.body.lg,
                                textAlign = TextAlign.Center,
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .testTag(testID?.let { "$it-searching-text" } ?: "intent-search-searching-text"),
                            )

                            // "This usually takes 1-2 seconds" (body.sm)
                            Text(
                                text = "This usually takes 1-2 seconds",
                                color = theme.colors.onSurface.default.copy(alpha = 0.6f),
                                style = theme.type.body.sm,
                                textAlign = TextAlign.Center,
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .testTag(testID?.let { "$it-searching-subtext" } ?: "intent-search-searching-subtext"),
                            )
                        }
                    }

                    is SearchState.OfflineUnsupported -> {
                        // Offline unsupported: empty state with recent intent chips
                        Column(
                            modifier = Modifier
                                .fillMaxWidth()
                                .testTag(testID?.let { "$it-offline" } ?: "intent-search-offline"),
                            horizontalAlignment = Alignment.CenterHorizontally,
                            verticalArrangement = Arrangement.spacedBy(theme.space.lg),
                        ) {
                            // Empty state message
                            Column(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .padding(horizontal = 24.dp),
                                horizontalAlignment = Alignment.CenterHorizontally,
                                verticalArrangement = Arrangement.spacedBy(theme.space.md),
                            ) {
                                // wifi-off icon (48dp)
                                IconSymbol(
                                    name = "wifi-off",
                                    size = 48.dp,
                                    color = theme.colors.onSurface.default.copy(alpha = 0.5f),
                                    modifier = Modifier
                                        .testTag(testID?.let { "$it-offline-icon" } ?: "intent-search-offline-icon"),
                                )

                                // "Connect to search" (title.md)
                                Text(
                                    text = "Connect to search",
                                    color = theme.colors.onSurface.default,
                                    style = theme.type.title.md,
                                    textAlign = TextAlign.Center,
                                    modifier = Modifier
                                        .fillMaxWidth()
                                        .testTag(testID?.let { "$it-offline-title" } ?: "intent-search-offline-title"),
                                )

                                // "You're offline. Try one of these popular rides instead:" (body.md)
                                Text(
                                    text = "You're offline. Try one of these popular rides instead:",
                                    color = theme.colors.onSurface.default.copy(alpha = 0.6f),
                                    style = theme.type.body.md,
                                    textAlign = TextAlign.Center,
                                    modifier = Modifier
                                        .fillMaxWidth()
                                        .testTag(testID?.let { "$it-offline-body" } ?: "intent-search-offline-body"),
                                )
                            }

                            // Recent intent chips (horizontal scroll)
                            LazyRow(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .padding(horizontal = theme.space.md)
                                    .testTag(testID?.let { "$it-recent-chips" } ?: "intent-search-recent-chips"),
                                horizontalArrangement = Arrangement.spacedBy(theme.space.sm),
                            ) {
                                items(searchState.recentIntents) { intent ->
                                    RecentIntentChip(
                                        intent = intent,
                                        onTap = onRecentIntentTap,
                                        modifier = Modifier
                                            .testTag("${testID?.let { "$it-" } ?: "intent-search-"}recent-chip-${intent.hashCode()}"),
                                    )
                                }
                            }
                        }
                    }

                    is SearchState.Results -> {
                        // Results: show intent summary pill
                        IntentSummaryPill(
                            summary = searchState.summary,
                            onDismiss = onClear,
                            modifier = Modifier
                                .testTag(testID?.let { "$it-results" } ?: "intent-search-results"),
                        )
                    }
                }
            }
        },
        confirmButton = { /* No confirm button for this sheet */ },
        dismissButton = { /* No dismiss button for this sheet */ },
        containerColor = theme.colors.surface.default,
        shape = RoundedCornerShape(theme.radius.xl),
    )
}

/**
 * IntentSummaryPill composable
 *
 * Simple pill showing intent summary with dismiss button.
 * Inline implementation since IntentSummaryPill atom doesn't exist yet on Android.
 *
 * @param summary Intent summary text to display
 * @param onDismiss Callback when dismiss button is tapped
 * @param modifier Modifier for the pill container
 */
@Composable
private fun IntentSummaryPill(
    summary: String,
    onDismiss: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val theme = LocalLaneShadowTheme.current

    Surface(
        modifier = modifier.fillMaxWidth(),
        shape = RoundedCornerShape(theme.radius.lg),
        color = theme.colors.surfaceVariant.default,
        border = BorderStroke(1.dp, theme.colors.border.default),
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(theme.space.md),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            // Summary text (body.md)
            Text(
                text = summary,
                color = theme.colors.onSurface.default,
                style = theme.type.body.md,
                modifier = Modifier.weight(1f),
            )

            // Dismiss button
            IconButton(
                onClick = onDismiss,
                modifier = Modifier.size(24.dp),
            ) {
                IconSymbol(
                    name = "close",
                    size = 18.dp,
                    color = theme.colors.onSurface.default.copy(alpha = 0.6f),
                )
            }
        }
    }
}

/**
 * RecentIntentChip composable
 *
 * Individual recent intent chip for offline state.
 *
 * @param intent Intent text to display
 * @param onTap Callback when chip is tapped
 * @param modifier Modifier for the chip
 */
@Composable
private fun RecentIntentChip(
    intent: String,
    onTap: (String) -> Unit,
    modifier: Modifier = Modifier,
) {
    val theme = LocalLaneShadowTheme.current

    Surface(
        modifier = modifier,
        shape = RoundedCornerShape(theme.radius.full),
        color = theme.colors.surfaceVariant.default,
        border = BorderStroke(1.dp, theme.colors.border.default),
        onClick = { onTap(intent) },
    ) {
        Text(
            text = intent,
            color = theme.colors.primary.default,
            style = theme.type.label.md,
            modifier = Modifier.padding(
                horizontal = 16.dp,
                vertical = 8.dp,
            ),
        )
    }
}

File: ios/LaneShadow/Views/Molecules/IntentSearchSheet.swift
import LaneShadowTheme
import SwiftUI

// MARK: - Search State Enum

/**
 * Search state for intent search sheet
 *
 * Represents the visual and interaction state of the search interface
 */
public enum LSSearchState: Sendable, Equatable {
    /// Empty input field ready for user input
    case idle
    /// Instant results with no spinner, shows intent summary pill
    case cacheHit(summary: String)
    /// Loading spinner with status message
    case searching
    /// Offline empty state with recent intent chips
    case offlineUnsupported(recentIntents: [String])
    /// Shows intent summary pill with results
    case results(summary: String)

    public static func == (lhs: LSSearchState, rhs: LSSearchState) -> Bool {
        switch (lhs, rhs) {
        case (.idle, .idle),
             (.searching, .searching):
            true
        case let (.cacheHit(lSummary), .cacheHit(rSummary)),
             let (.results(lSummary), .results(rSummary)):
            lSummary == rSummary
        case let (.offlineUnsupported(lIntents), .offlineUnsupported(rIntents)):
            lIntents == rIntents
        default:
            false
        }
    }
}

// MARK: - Intent Search Sheet Component

/**
 * Intent search sheet molecule component
 *
 * Bottom sheet for natural language route discovery (UC-DISC-07).
 * Displays four visual states: idle, cache_hit, searching, offline_unsupported.
 * Following React Native component from react-native/components/discovery/intent-search-sheet.tsx
 *
 * ## Design Tokens Used
 * - Colors:
 *   - Background: `theme.colors.surface.default`
 *   - Border: `theme.colors.border.default`
 *   - Header text: `theme.colors.onSurface.default`, `theme.colors.onSurface.muted`
 *   - Input row background: `theme.colors.surface.default`
 *   - Input row border: `theme.colors.border.default`
 *   - Search icon: `theme.colors.onSurface.muted`
 *   - Text: `theme.colors.onSurface.default`, `theme.colors.onSurface.subtle`
 *   - Primary: `theme.colors.primary.default`
 *   - Surface variant: `theme.colors.surfaceVariant.default`
 * - Layout:
 *   - Container gap: `theme.space.md` (16pt)
 *   - Header gap: `theme.space.xs` (4pt)
 *   - Header padding bottom: `theme.space.lg` (16pt)
 *   - Input row height: 48pt
 *   - Input row padding: `theme.space.md` (12pt)
 *   - Icon margin: `theme.space.sm` (8pt)
 *   - Clear button: 32x32pt
 *   - State container gap: `theme.space.md` to `theme.space.lg`
 *   - Chip padding: `theme.space.md` horizontal, `theme.space.sm` vertical
 *   - Chip radius: `theme.radius.full`
 * - Typography:
 *   - Title: `theme.type.title.lg`
 *   - Subtitle: `theme.type.body.md`
 *   - Input text: `theme.type.body.lg`
 *   - Status text: `theme.type.body.lg`, `theme.type.body.sm`
 *   - Offline title: `theme.type.title.md`
 *   - Offline message: `theme.type.body.md`
 *   - Chip text: `theme.type.label.md`
 * - Radius:
 *   - Input row: `theme.radius.lg`
 *   - Chips: `theme.radius.full`
 * - States:
 *   - idle: Empty input with placeholder
 *   - cache_hit: Shows LSIntentSummaryPill
 *   - searching: ProgressView with status messages
 *   - offline_unsupported: Empty state with wifi-off icon and chips
 *   - results: Shows LSIntentSummaryPill
 *
 * ## Parameters
 * - searchState: Current search state (determines which UI to render)
 * - onSearch: Callback when user submits search query
 * - onClear: Callback when user taps clear button (resets to browse mode)
 * - onRecentIntentTap: Callback when user taps a recent-intent chip in offline state
 * - visible: Whether the sheet is visible
 * - query: Current search query text
 * - onChangeQuery: Callback when query text changes
 * - testID: Optional testing identifier for UI tests
 *
 * ## Accessibility
 * - Header: "Describe your ideal ride" label
 * - Input: "Search for rides" label
 * - Clear button: "Clear search" label
 * - Recent intent chips: "{intent} ride" label each
 */
public struct LSIntentSearchSheet: View {
    @Environment(\.theme) private var theme

    private let searchState: LSSearchState
    private let onSearch: (String) -> Void
    private let onClear: () -> Void
    private let onRecentIntentTap: (String) -> Void
    private let visible: Bool
    @Binding private var query: String
    private let onChangeQuery: (String) -> Void
    private let testID: String

    public init(
        searchState: LSSearchState,
        onSearch: @escaping (String) -> Void,
        onClear: @escaping () -> Void,
        onRecentIntentTap: @escaping (String) -> Void,
        visible: Bool,
        query: Binding<String>,
        onChangeQuery: @escaping (String) -> Void,
        testID: String = "intent-search-sheet"
    ) {
        self.searchState = searchState
        self.onSearch = onSearch
        self.onClear = onClear
        self.onRecentIntentTap = onRecentIntentTap
        self.visible = visible
        _query = query
        self.onChangeQuery = onChangeQuery
        self.testID = testID
    }

    // MARK: - Body

    public var body: some View {
        VStack(spacing: 0) {
            if visible {
                sheetContent
            }
        }
    }

    // MARK: - Sheet Content

    private var sheetContent: some View {
        VStack(spacing: theme.space.md) {
            // Header
            headerView

            // Input row with clear button
            inputRowView

            // State-specific content
            stateContentView
        }
        .padding(theme.space.lg)
        .background(theme.colors.surface.default)
        .accessibilityIdentifier(testID)
    }

    // MARK: - Header

    private var headerView: some View {
        VStack(alignment: .leading, spacing: theme.space.xs) {
            Text("Describe your ideal ride")
                .font(theme.type.title.lg.font)
                .foregroundStyle(theme.colors.onSurface.default)

            Text("Say \"scenic coastal roads\" or \"twisty mountain passes\"")
                .font(theme.type.body.md.font)
                .foregroundStyle(theme.colors.onSurface.muted)
        }
        .padding(.bottom, theme.space.lg)
        .overlay(
            Rectangle()
                .fill(theme.colors.border.default)
                .frame(height: 1),
            alignment: .bottom
        )
        .accessibilityElement(children: .combine)
        .accessibilityLabel("Describe your ideal ride search")
    }

    // MARK: - Input Row

    private var inputRowView: some View {
        HStack(alignment: .center, spacing: 0) {
            // Search icon and text input
            HStack(alignment: .center, spacing: theme.space.sm) {
                LSIconSymbol(
                    name: "magnify",
                    size: 20,
                    color: theme.colors.onSurface.muted
                )
                .padding(.leading, 4)

                Text(query.isEmpty ? "Type your ideal ride..." : query)
                    .font(theme.type.body.lg.font)
                    .foregroundStyle(
                        query.isEmpty
                            ? theme.colors.onSurface.subtle
                            : theme.colors.onSurface.default
                    )
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .onTapGesture {
                        // In production, this would focus a TextField
                        // For design mock, just show the query text
                    }
            }

            // Clear button
            if !query.isEmpty || isSearching {
                Button(action: onClear) {
                    LSIconSymbol(
                        name: "close",
                        size: 20,
                        color: theme.colors.onSurface.muted
                    )
                    .frame(width: 32, height: 32)
                    .contentShape(Rectangle())
                }
                .buttonStyle(PlainButtonStyle())
                .accessibilityLabel("Clear search")
                .accessibilityIdentifier("\(testID)-clear")
            }
        }
        .frame(height: 48)
        .padding(.horizontal, 12)
        .background(theme.colors.surface.default)
        .clipShape(RoundedRectangle(cornerRadius: theme.radius.lg))
        .overlay(
            RoundedRectangle(cornerRadius: theme.radius.lg)
                .stroke(theme.colors.border.default, lineWidth: 1)
        )
        .accessibilityElement(children: .contain)
        .accessibilityLabel("Search input")
    }

    // MARK: - State Content

    @ViewBuilder
    private var stateContentView: some View {
        switch searchState {
        case .idle:
            idleView

        case let .cacheHit(summary):
            cacheHitView(summary: summary)

        case .searching:
            searchingView

        case let .offlineUnsupported(recentIntents):
            offlineView(recentIntents: recentIntents)

        case let .results(summary):
            resultsView(summary: summary)
        }
    }

    // MARK: - Idle View

    private var idleView: some View {
        VStack(spacing: theme.space.md) {
            Spacer()
                .frame(minHeight: 100)
        }
        .accessibilityIdentifier("\(testID)-idle")
    }

    // MARK: - Cache Hit View

    private func cacheHitView(summary: String) -> some View {
        VStack(spacing: theme.space.md) {
            LSIntentSummaryPill(
                text: summary,
                onDismiss: onClear,
                testID: "\(testID)-cache-pill"
            )

            Spacer()
                .frame(minHeight: 100)
        }
        .accessibilityIdentifier("\(testID)-cache-hit")
    }

    // MARK: - Searching View

    private var searchingView: some View {
        VStack(spacing: theme.space.md) {
            Spacer()
                .frame(minHeight: 40)

            ProgressView()
                .scaleEffect(1.2)
                .tint(theme.colors.primary.default)
                .accessibilityIdentifier("\(testID)-spinner")

            Text("Finding your perfect ride...")
                .font(theme.type.body.lg.font)
                .foregroundStyle(theme.colors.onSurface.default)

            Text("This usually takes 1-2 seconds")
                .font(theme.type.body.sm.font)
                .foregroundStyle(theme.colors.onSurface.muted)

            Spacer()
                .frame(minHeight: 60)
        }
        .accessibilityIdentifier("\(testID)-searching")
    }

    // MARK: - Offline View

    private func offlineView(recentIntents: [String]) -> some View {
        VStack(spacing: theme.space.lg) {
            // Empty state message
            VStack(spacing: theme.space.md) {
                LSIconSymbol(
                    name: "wifi-off",
                    size: 48,
                    color: theme.colors.onSurface.subtle
                )
                .accessibilityIdentifier("\(testID)-offline-icon")

                Text("Connect to search")
                    .font(theme.type.title.md.font)
                    .foregroundStyle(theme.colors.onSurface.default)

                Text("You're offline. Try one of these popular rides instead:")
                    .font(theme.type.body.md.font)
                    .foregroundStyle(theme.colors.onSurface.muted)
                    .multilineTextAlignment(.center)
            }

            // Recent intent chips
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: theme.space.sm) {
                    ForEach(Array(recentIntents.enumerated()), id: \.offset) { index, intent in
                        Button(action: { onRecentIntentTap(intent) }) {
                            Text(intent)
                                .font(theme.type.label.md.font)
                                .foregroundStyle(theme.colors.primary.default)
                                .padding(.horizontal, theme.space.md)
                                .padding(.vertical, theme.space.sm)
                                .background(theme.colors.surfaceVariant.default)
                                .clipShape(Capsule())
                                .overlay(
                                    Capsule()
                                        .stroke(theme.colors.border.default, lineWidth: 1)
                                )
                        }
                        .buttonStyle(PlainButtonStyle())
                        .accessibilityLabel("\(intent) ride")
                        .accessibilityIdentifier("\(testID)-chip-\(index)")
                    }
                }
                .padding(.horizontal, theme.space.md)
            }
        }
        .accessibilityIdentifier("\(testID)-offline")
    }

    // MARK: - Results View

    private func resultsView(summary: String) -> some View {
        VStack(spacing: theme.space.md) {
            LSIntentSummaryPill(
                text: summary,
                onDismiss: onClear,
                testID: "\(testID)-results-pill"
            )

            Spacer()
                .frame(minHeight: 100)
        }
        .accessibilityIdentifier("\(testID)-results")
    }

    // MARK: - Computed Properties

    private var isSearching: Bool {
        if case .searching = searchState {
            return true
        }
        return false
    }
}

// MARK: - Preview

#Preview("Intent Search Sheet - Idle") {
    struct IdleDemo: View {
        @State private var query = ""

        var body: some View {
            LSIntentSearchSheet(
                searchState: .idle,
                onSearch: { _ in },
                onClear: {},
                onRecentIntentTap: { _ in },
                visible: true,
                query: $query,
                onChangeQuery: { _ in }
            )
            .laneShadowTheme()
        }
    }

    return IdleDemo()
}

#Preview("Intent Search Sheet - Cache Hit") {
    struct CacheHitDemo: View {
        @State private var query = "scenic coastal roads"

        var body: some View {
            LSIntentSearchSheet(
                searchState: .cacheHit(summary: "Twisty mountain roads near you"),
                onSearch: { _ in },
                onClear: {},
                onRecentIntentTap: { _ in },
                visible: true,
                query: $query,
                onChangeQuery: { _ in }
            )
            .laneShadowTheme()
        }
    }

    return CacheHitDemo()
}

#Preview("Intent Search Sheet - Searching") {
    struct SearchingDemo: View {
        @State private var query = "twisty mountain passes"

        var body: some View {
            LSIntentSearchSheet(
                searchState: .searching,
                onSearch: { _ in },
                onClear: {},
                onRecentIntentTap: { _ in },
                visible: true,
                query: $query,
                onChangeQuery: { _ in }
            )
            .laneShadowTheme()
        }
    }

    return SearchingDemo()
}

#Preview("Intent Search Sheet - Offline") {
    struct OfflineDemo: View {
        @State private var query = ""

        var body: some View {
            LSIntentSearchSheet(
                searchState: .offlineUnsupported(recentIntents: [
                    "Scenic coastal routes",
                    "Twisty mountain passes",
                    "Wine country backroads",
                    "Desert sunset rides",
                ]),
                onSearch: { _ in },
                onClear: {},
                onRecentIntentTap: { print("Tapped: \($0)") },
                visible: true,
                query: $query,
                onChangeQuery: { _ in }
            )
            .laneShadowTheme()
        }
    }

    return OfflineDemo()
}

#Preview("Intent Search Sheet - Results") {
    struct ResultsDemo: View {
        @State private var query = "scenic routes near me"

        var body: some View {
            LSIntentSearchSheet(
                searchState: .results(summary: "Scenic coastal routes with elevation gain"),
                onSearch: { _ in },
                onClear: {},
                onRecentIntentTap: { _ in },
                visible: true,
                query: $query,
                onChangeQuery: { _ in }
            )
            .laneShadowTheme()
        }
    }

    return ResultsDemo()
}

#Preview("Intent Search Sheet - All States") {
    struct AllStatesDemo: View {
        @State private var currentState: LSSearchState = .idle
        @State private var query = ""

        var body: some View {
            VStack(spacing: 20) {
                // State picker
                Picker("State", selection: $currentState) {
                    Text("Idle").tag(LSSearchState.idle)
                    Text("Cache Hit").tag(LSSearchState.cacheHit(summary: "Twisty mountain roads"))
                    Text("Searching").tag(LSSearchState.searching)
                    Text("Offline").tag(LSSearchState.offlineUnsupported(recentIntents: [
                        "Scenic routes",
                        "Mountain passes",
                    ]))
                    Text("Results").tag(LSSearchState.results(summary: "Scenic coastal routes"))
                }
                .pickerStyle(.segmented)
                .padding()

                // Sheet
                LSIntentSearchSheet(
                    searchState: currentState,
                    onSearch: { _ in },
                    onClear: { query = "" },
                    onRecentIntentTap: { _ in },
                    visible: true,
                    query: $query,
                    onChangeQuery: { _ in }
                )
            }
            .laneShadowTheme()
        }
    }

    return AllStatesDemo()
}
